subsys/net/ip/icmp.c

```diff
@@ -0,0 +1,479 @@
+/** @file
+ * @brief ICMP related functions
+ */
+
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Use highest log level if both IPv4 and IPv6 are defined */
+#if defined(CONFIG_NET_IPV4)
+#define MY_LOG_LEVEL CONFIG_NET_ICMPV4_LOG_LEVEL
+#endif
+
+#if defined(CONFIG_NET_IPV6)
+#if defined(MY_LOG_LEVEL)
+#if MY_LOG_LEVEL < CONFIG_NET_ICMPV6_LOG_LEVEL
+#undef MY_LOG_LEVEL
+#define MY_LOG_LEVEL CONFIG_NET_ICMPV6_LOG_LEVEL
+#endif
+#else
+#define MY_LOG_LEVEL CONFIG_NET_ICMPV6_LOG_LEVEL
+#endif
+#endif
+
+#if !defined(MY_LOG_LEVEL)
+#define MY_LOG_LEVEL LOG_LEVEL_INF
+#endif
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(net_icmp, MY_LOG_LEVEL);
+
+#include <errno.h>
+#include <zephyr/random/rand32.h>
+#include <zephyr/sys/slist.h>
+#include <zephyr/net/net_pkt.h>
+#include <zephyr/net/icmp.h>
+
+#include "net_private.h"
+#include "icmpv6.h"
+#include "icmpv4.h"
+#include "ipv4.h"
+#include "ipv6.h"
+#include "net_stats.h"
+
+static K_MUTEX_DEFINE(lock);
+static sys_slist_t handlers;
+
+#define PKT_WAIT_TIME K_SECONDS(1)
+
+int net_icmp_init_ctx(struct net_icmp_ctx *ctx, uint8_t type, uint8_t code,
+		      net_icmp_handler_t handler)
+{
+	if (ctx == NULL || handler == NULL) {
+		return -EINVAL;
+	}
+
+	memset(ctx, 0, sizeof(struct net_icmp_ctx));
+
+	ctx->handler = handler;
+	ctx->type = type;
+	ctx->code = code;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	sys_slist_prepend(&handlers, &ctx->node);
+
+	k_mutex_unlock(&lock);
+
+	return 0;
+}
+
+int net_icmp_cleanup_ctx(struct net_icmp_ctx *ctx)
+{
+	if (ctx == NULL) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	sys_slist_find_and_remove(&handlers, &ctx->node);
+
+	k_mutex_unlock(&lock);
+
+	memset(ctx, 0, sizeof(struct net_icmp_ctx));
+
+	return 0;
+}
+
+#if defined(CONFIG_NET_IPV4)
+static int send_icmpv4_echo_request(struct net_icmp_ctx *ctx,
+				    struct net_if *iface,
+				    struct in_addr *dst,
+				    uint16_t identifier,
+				    uint16_t sequence,
+				    uint8_t tos,
+				    int priority,
+				    const void *data,
+				    size_t data_size,
```

Comment:
Maybe this internal function could accept `struct net_icmp_ping_params*` as well?

Reply:
Sure, np.